sdk/typescript/telemetry/telemetry.ts

```diff
@@ -0,0 +1,222 @@
+   */
+  async close() {
+    if (this.sdk) {
+      this.sdk.shutdown()
+    }
+  }
+
+  configure() {
+    if (!otelConfigured()) {
+      console.debug("Telemetry not configured")
+      return
+    }
+
+    if (!otelEnabled()) {
+      console.debug("Telemetry disabled")
+      return
+    }
+
+    this.setupEnv()
+  }
+
+  /**
+   * Setup environment for auto-configuring the SDK.
+   */
+  setupEnv() {
+    // Setup a default value for the service name.
+    if (!process.env[OtelEnv.OTEL_SERVICE_NAME]) {
+      process.env[OtelEnv.OTEL_SERVICE_NAME] = SERVICE_NAME
+    }
+
+    // TODO: The insecure flag should be set by the shim instead of the SDK.
+    const endpoints = [
+      OtelEnv.OTEL_EXPORTER_OTLP_ENDPOINT,
+      OtelEnv.OTEL_EXPORTER_OTLP_METRICS_ENDPOINT,
+      OtelEnv.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT,
+      OtelEnv.OTEL_EXPORTER_OTLP_TRACES_ENDPOINT,
+    ]
+
+    endpoints.forEach((endpoint) => {
+      const env = process.env[endpoint]
+      if (env && env.startsWith("http://")) {
+        process.env[OtelEnv.OTEL_EXPORTER_OTLP_INSECURE] = "true"
+      }
+    })
+  }
+}
+
+const configurator = new DaggerOtelConfigurator()
+
+export function initiliaze() {
+  configurator.initialize()
+}
+
+export async function close() {
+  await configurator.close()
+}
+
+/**
+ * Return a tracer to use with Dagger.
+ *
+ * The tracer is automatically initialized if not already done.
+ * As a conveniance function, you can use `withTracingSpan` that automatically close
+ * the span at the end of the function.
+ */
+export function getTracer(): opentelemetry.Tracer {
+  initiliaze()
+
+  const tracer = opentelemetry.trace
+    .getTracerProvider()
+    .getTracer("dagger-typescript-sdk")
+
+  return tracer
+}
+
+export function initContext() {
+  const ctx = opentelemetry.context.active()
+
+  const parentID = process.env.TRACEPARENT
+  if (parentID) {
+    return opentelemetry.propagation.extract(ctx, {
+      traceparent: parentID,
+    })
+  }
+
+  return ctx
+}
+
+/**
+ * Execute the functions with a custom span with the given name using startActiveSpan.
+ * The function executed will use the parent context of the function (it can be another span
+ * or the main function).
+ *
+ * @param name The name of the span
+ * @param fn The functions to execute
+ *
+ * WithTracingSpan returns the result of the executed functions.
+ *
+ * The span is automatically ended when the function is done.
+ * The span is automatically marked as an error if the function throws an error.
+ *
+ *
+ * @example
+ * ```
+ * return withTracingSpan(name, async () => {
+ *   return this.containerEcho("test").stdout()
+ * })
+ * ```
+ */
+export async function withTracingSpan<T>(
+  name: string,
+  fn: (span: opentelemetry.Span) => Promise<T>,
+): Promise<T> {
+  return await opentelemetry.context.with(
+    opentelemetry.context.active(),
+    async () => {
+      return getTracer().startActiveSpan(name, async (span) => {
```

Comment:
I'm not sure about `getTracer()` here because it hardcodes the tracer name. That's all good if it's used inside the SDK, but when users are doing it in their own functions, they should use their own name. 

Python and Go's equivalent of `getTracer` is meant to be used by the SDK, not by users. In Go and Python users should use normal opentelemetry code (to get the tracer and start the span). It's just the configuration that's handled for them.

In Go, you need to initialize manually (`telemetry.InitEmbeded()`), but Python and Typescript is doing that over the connection so it's not important for user spans to properly initialize unless it's outside of a dagger connection and that doesn't happen on modules, only in `dagger run`.

You could either:
- have a `name` argument in `getTracer` that defaults to the SDK's and pass it on here somehow
- let user give the Tracer to `withTracingSpan`
- wrap this in a class so users can do `getTracer("my-module").startActiveSpan()`, where `startActiveSpan` replaces `withTracingSpan`. The same name brings familiarity, you'd just be handling the catch/finally branches.

Reply:
Added, I wrapped the tracer in a Tracer class